ipykernel/iostream.py

```diff
@@ -26,76 +31,75 @@
-class OutStream(object):
-    """A file like object that publishes the stream to a 0MQ PUB socket."""
-
-    # The time interval between automatic flushes, in seconds.
-    _subprocess_flush_limit = 256
-    flush_interval = 0.05
-    topic=None
-
-    def __init__(self, session, pub_socket, name, pipe=True):
-        self.encoding = 'UTF-8'
-        self.session = session
-        self.pub_socket = pub_socket
-        self.name = name
-        self.topic = b'stream.' + py3compat.cast_bytes(name)
-        self.parent_header = {}
-        self._new_buffer()
-        self._buffer_lock = threading.Lock()
+class IOPubThread(object):
+    """An object for sending IOPub messages in a background thread
+    
+    prevents a blocking main thread
+    
+    IOPubThread(pub_socket).background_socket is a Socket-API-providing object
+    whose IO is always run in a thread.
+    """
+
+    def __init__(self, socket, pipe=False):
+        self.socket = socket
+        self.background_socket = BackgroundSocket(self)
         self._master_pid = os.getpid()
-        self._master_thread = threading.current_thread().ident
         self._pipe_pid = os.getpid()
         self._pipe_flag = pipe
+        self.io_loop = IOLoop()
         if pipe:
             self._setup_pipe_in()
+        self.thread = threading.Thread(target=self._thread_main)
+        self.thread.daemon = True
+
+    def _thread_main(self):
+        """The inner loop that's actually run in a thread"""
+        self.io_loop.start()
+        self.io_loop.close()
 
     def _setup_pipe_in(self):
         """setup listening pipe for subprocesses"""
-        ctx = self.pub_socket.context
+        ctx = self.socket.context
 
         # use UUID to authenticate pipe messages
         self._pipe_uuid = uuid.uuid4().bytes
 
-        self._pipe_in = ctx.socket(zmq.PULL)
-        self._pipe_in.linger = 0
+        pipe_in = ctx.socket(zmq.PULL)
+        pipe_in.linger = 0
         try:
-            self._pipe_port = self._pipe_in.bind_to_random_port("tcp://127.0.0.1")
+            self._pipe_port = pipe_in.bind_to_random_port("tcp://127.0.0.1")
         except zmq.ZMQError as e:
-            warn("Couldn't bind IOStream to 127.0.0.1: %s" % e +
+            warn("Couldn't bind IOPub Pipe to 127.0.0.1: %s" % e +
                 "\nsubprocess output will be unavailable."
             )
             self._pipe_flag = False
-            self._pipe_in.close()
-            del self._pipe_in
+            pipe_in.close()
             return
-        self._pipe_poller = zmq.Poller()
-        self._pipe_poller.register(self._pipe_in, zmq.POLLIN)
-        if IOLoop.initialized():
-            # subprocess flush should trigger flush
-            # if kernel is idle
-            IOLoop.instance().add_handler(self._pipe_in,
-                lambda s, event: self.flush(),
-                IOLoop.READ,
-            )
+        self._pipe_in = ZMQStream(pipe_in, self.io_loop)
+        self._pipe_in.on_recv(self._handle_pipe_msg)
```

Comment:
This is the norm for pipes, I think. Messages *go out* on pipe_out (like stdout) and *come in* on pipe_in (like stdin); pipe_out leads out of its process, pipe_in leads into its process.